### PR TITLE
Add test-design skill (古典技法 / Hoare論理・DbC / PBT / Google Test Sizes)

### PR DESCRIPTION
**Pull request by kanade0404**
## 概要

テストを**書く前**に「何をテストすべきか」を体系的に導出する新 skill **test-design** を追加する。出力は固定フォーマットの「テストケース設計表」で、各行が `tdd` Step 1 の RED テスト 1 本に対応する。

## 背景

過去セッション・PR レビュー指摘の採掘 (#28 参照) と既存 skill 棚卸しの結果、次の 2 領域がハーネスに**完全に欠落**していた:

- 古典的テストケース設計技法 (同値分割 / 境界値 / デシジョンテーブル / 状態遷移 / ペアワイズ) — grep 0 件
- Hoare 論理 / Design by Contract による契約からの assertion 導出 — grep 0 件

また test-review の coverage-gap 指摘 (境界ケース欠落等) に「解消先」となる skill が無く、boundary-gap 系の指摘が放置される構造だった (CodeRabbit が skills PR 1/3/9/14 で同型の境界ケース欠落を繰り返し指摘)。

## 内容

- `skills/test-design/SKILL.md` (155 行) — 契約明文化 → 技法選択決定表 → ケース列挙 → Test Sizes 配分 → 設計表出力の 5 Step。トレース不能な仮定は「未決の契約質問」に分離し期待値を発明しない規律を含む
- `references/classical-techniques.md` — 同値分割 / BVA (on/off point) / デシジョンテーブル / 状態遷移 (0-switch/1-switch の既定基準付き) / ペアワイズ
- `references/contracts-and-properties.md` — Hoare 論理 ({P} C {Q}) からのテスト導出 + PBT property taxonomy (invariant / roundtrip / oracle / metamorphic / idempotence)。ツール詳細は test-review/references/python.md と重複させない
- `references/test-sizes.md` — Google Software Engineering の Small/Medium/Large 配分、change-detector test 回避、hermetic 設計

## 責務境界

| 隣接 skill | 境界 |
|---|---|
| test-review | 既存テストのレビュー (本 skill は書く前の設計) |
| tdd | RED-GREEN 実行 (本 skill は Step 1 の前段に設計表を渡す) |
| test-mutation-gate (#28) | テスト diff の検出力監査 (boundary-gap warn の解消先が本 skill) |

## 実験・評価 (実施済み)

- **Trigger eval**: 17 ケース (explicit 5 / ambiguous 4 / adjacent 5 / distractor 3)。作成者と別の独立 judge にラベル非開示で判定させ **F1 = 1.0** (`evals/test-design-trigger-results-2026-07-05.jsonl`)
- **品質 eval (skill-builder Mode C)**: 白紙 subagent による 2 iteration
  - iter 1 (クーポン適用関数、部分仕様): checklist 7/7 (critical 2 含む)、不明瞭点 5 件 → General Fix Rule を全て適用
  - iter 2 (状態遷移 + roundtrip、未記載遷移あり): checklist 7/7、iter 1 の修正が機能 (「未決の契約質問」経路が正しく使用された)。新規不明瞭点 3 件 (磨き込みレベル) → 適用済み
  - 収束傾向 (5→3、critical 0) を確認してループ終了。iter 3 での再検証は未実施 (今後の運用で回収)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019HJPNWS3MTB7ggCvhnKRow

## Summary by CodeRabbit

* **New Features**
  * Added a new test-design workflow that generates a standardized “テストケース設計表” before test implementation.
  * Automated guidance for selecting techniques, enumerating cases, and assigning Test Size, including contract- and property-based scenarios.
* **Documentation**
  * Added/expanded reference materials for classical test techniques, contract/property-driven test derivation, and Test Size assignment rules.
* **Tests**
  * Added evaluation configurations and expected trigger outcomes to improve consistency of test-design skill detection.

### CODE REVIEW

File: skills/test-design/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+- 「テストケース考えて」「どんなテスト書けばいい?」「テスト設計して」
+- 「テスト観点洗い出して」「境界値どこ?」「property 何にする?」
+- 「この関数の事前条件・事後条件は?」「不変条件は何?」
+- 「実装する前にテストケース洗い出したい」「仕様からテスト導出して」
+- `tdd` Step 1 (RED) に入る前、書くべきテストの一覧が定まっていないとき
+
+**発火しない場面** → 該当 skill 名を明示する:
+- 既に書かれたテストコードの品質・smell・seam を見る → `test-review`
+- 設計表の 1 行を実際に RED テストとして書き、実装して GREEN にする → `tdd`
+- 書かれたテストが実装のミューテーションを検出できるかを監査する（boundary-gap warn の解消） → `test-mutation-gate`
```
[the remark this answers is not in the document]
Comment (kanade0404) [reply skills/test-design/SKILL.md:34]: ご指摘の通り、現時点で `skills/test-mutation-gate/` には SKILL.md が存在せず（master にも `feat/test-design` にも未マージ）、この PR の負の空間ルーティング（「テスト diff の検出力監査 → `test-mutation-gate`」）は一時的にダングリング参照になります。

ただしこの skill 自体は既に別 PR で開発が進行中です（#28「Add test-mutation-gate skill (Phase 1)」、#35 がその上にスタック）。本 PR で `test-mutation-gate` を新規に作り込む、または #28 のマージを待って本 PR の順序を調整するのは、本 PR 単体のスコープ外と判断しました。

Tracked in #28. #28 がマージされ次第この参照は解消されます。

File: skills/test-design/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
+---
+name: test-design
```
[the remark this answers is not in the document]
Comment (kanade0404) [reply skills/test-design/SKILL.md:2]: Fixed in 1e7727d. Ready for re-review.

`node scripts/rulesync-sync.mjs` 相当の generate を実行し、`.claude/skills/test-design/` と `.agents/skills/test-design/` を追加済みです（このコメントの直後のコミットで対応済みでした）。念のためこのセッションでも drift チェック相当を再実行し、`skills/test-design` と生成物 (`.claude`, `.agents`) の間に差分が無いことを確認しました。

File: skills/test-design/references/contracts-and-properties.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+### Idempotence（冪等性）
+`f(f(x)) == f(x)`。同じ操作を 2 回適用しても結果が変わらない性質。
+例: 「正規化関数を 2 回適用しても 1 回適用と同じ」「PUT リクエストを 2 回送っても状態は 1 回目と同じ」
+
+**shrinking しやすい property の書き方.**
+- property の前提条件（`assume`/`pre` のようなフィルタ）を最小限にする。フィルタが厳しいほど shrinking が失敗しやすい。
+- 生成する入力の構造をシンプルに保つ（ネストの深いオブジェクトより、フラットな構造の組合せで表現できないか検討する）。
+- 失敗時に人間が読める最小反例が得られるよう、生成戦略はまず単純な型（整数・短い文字列）から試す。
+
+具体的なツール（Hypothesis の `@given`/`strategies`、fast-check の `fc.assert`/`fc.property` 等）の書き方は `skills/test-review/references/python.md` を参照する。ここでは重複させない。
```
[the remark this answers is not in the document]
Comment (kanade0404) [reply skills/test-design/references/contracts-and-properties.md]: Fixed in 99cc22a. Ready for re-review.

ご指摘の通り `skills/test-review/references/python.md` は Python/Hypothesis 専用で、fast-check (JS/TS) の詳細はありませんでした。fast-check への言及を「python.md を参照」から「本カタログに専用リファレンスが無いため各ツールの公式ドキュメントを参照する」に修正しました（`contracts-and-properties.md` 2 箇所、`SKILL.md` の非ゴール節 1 箇所）。

File: skills/test-design/references/test-sizes.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+分類の基準は「テストが何に依存して落ちうるか」であって、テスト対象のコード量ではない。1 行の関数でもネットワーク越しに実行すれば Large になり、100 行のロジックでも純粋な入出力変換なら Small になる。
+
+---
+
+## どの振る舞いをどのサイズに置くか
+
+Step 4 でケースごとに判断する基準:
+
+1. **外部 I/O があるか** — DB・ファイル・ネットワーク・外部プロセスに触れるなら Small から外れる。
+2. **時間・並行性に依存するか** — 実クロックや実タイムアウトに依存するなら Medium/Large。決定的な Clock を注入できるなら Small のままにできる（`test-review` §6 の Clock DI 参照）。ただし待機・ポーリングを伴わない一回性の時刻比較（有効期限判定など）で、閾値に実運用上十分な余裕がある場合は Small のままでよい。
```
[the remark this answers is not in the document]
Comment (kanade0404) [reply skills/test-design/references/test-sizes.md]: Fixed in 99cc22a. Ready for re-review.

「待機・ポーリングを伴わない一回性の時刻比較なら、閾値に余裕があれば real clock のまま Small でよい」という例外は、同じ reference が別所で要求する「決定的な入力生成: 現在時刻を使うなら seed/注入可能な Clock を前提にする」という hermetic の原則と矛盾していました。この例外を削除し、Small に置くなら常に注入可能な Clock を使う（閾値の余裕だけでは hermetic 性を代替できない）と明記しました。

File: skills/test-design/references/contracts-and-properties.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+## 1. 契約からの導出 (Hoare 論理 / Design by Contract)
+
+**表記.** `{P} C {Q}` — 事前条件 `P` を満たす状態でコマンド `C` を実行すると、事後条件 `Q` を満たす状態で終了する。加えて `C` の実行を通じて常に成り立つべき **不変条件 (invariant)** を持つ場合がある。
+
+**手順.**
+1. 対象のシグネチャから `P`（呼び出し側が保証すべき前提）を全て書き出す。型だけで表現できない制約（「配列は空でない」「日付は未来でない」等）を優先的に拾う。
+2. `Q`（呼び出し後に保証される性質）を書き出す。戻り値の性質だけでなく、副作用（状態変化、外部への書き込み）も含める。
+3. `invariant`（呼び出し前後で常に成り立つ性質、例: 「リストの長さは減らない」「合計金額は負にならない」）を書き出す。無ければ「なし」と明記する。
+4. 契約設計の判断: **defensive contract**（`P` 違反時に例外/エラーで防御する）か **wide contract**（`P` の範囲を広げて呼び出し側の制約を減らす）かを決める。この判断自体が設計判断であり、テスト設計はその判断の結果に従う（defensive なら (b) のエラー系テストを書く、wide なら該当入力を正常系として (a) に含める）。
```
[the remark this answers is not in the document]
Comment (kanade0404) [reply skills/test-design/references/contracts-and-properties.md]: Fixed in 99cc22a. Ready for re-review.

defensive/wide contract の選択がテスト設計スキル自身の判断であるかのように読める記述になっていました。これは SKILL.md が明示する「API 仕様そのものを決めるのは design/software-design」という境界、および Step 1 の「トレースできない前提から黙って推測しない」規律と矛盾します。既存仕様からトレース可能な場合はそれに従い、トレースできない場合は `AskUserQuestion` で確認するか「要確認」タグを付けて design/software-design に委ねる、と明記しました。

File: skills/test-design/references/classical-techniques.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+---
+
+## 5. ペアワイズ / 組合せテスト (Pairwise / Combinatorial Testing)
+
+**定義.** パラメータ数が多く全組合せ (`n1 × n2 × ... × nk`) が非現実的な場合、実務上の欠陥の大半は 2 パラメータ間の相互作用で発生するという経験則に基づき、全ペアが最低 1 回は出現する縮小組合せでカバーする。
+
+**手順.**
+1. 各パラメータの水準（値の代表クラス、同値分割/BVA で既に絞ったもの）を列挙する。
+2. 全パラメータの「任意の 2 つの組合せ」が最低 1 回出現する最小の組合せ集合を作る（直交表やペアワイズ生成アルゴリズムに委ねる。具体的なツール実行は本スキルの範囲外 — 生成ロジック自体は実行者やツールに任せる）。
```
[the remark this answers is not in the document]
Comment (kanade0404) [reply skills/test-design/references/classical-techniques.md]: Fixed in 99cc22a. Ready for re-review.

ペアワイズの水準列挙が「同値分割/BVA で絞ったもの」を無条件に含めていたため、無効クラスの代表値も組合せに混ざり得ました。これは §1 の「無効クラスは 1 パラメータにつき 1 つずつ独立にテストする」という原則と矛盾します。ペアワイズの組合せ集合に含めるのは有効クラスの代表値のみとし、無効クラスは §1 の原則どおり別枠のケースとして追加する、と明記しました。

File: skills/test-design/references/classical-techniques.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+## 2. 境界値分析 (Boundary Value Analysis, BVA)
+
+**定義.** 欠陥は境界の近くに集中するという経験則に基づき、各境界の **on-point**（境界値そのもの）と **off-point**（境界のすぐ外側）を両方テストする。
+
+**手順.**
+1. Step 1 で明文化した範囲・長さ・件数の境界を全て列挙する（最小値・最大値・0・空・上限件数など）。
+2. 各境界について次の値を候補にする:
+   - **on-point**: 境界値そのもの（`min`, `max`）
+   - **off-point**: 境界から ±1（`min - 1`, `min + 1`, `max - 1`, `max + 1`）
```
[the remark this answers is not in the document]
Comment (kanade0404) [reply skills/test-design/references/classical-techniques.md]: Fixed in 99cc22a. Ready for re-review.

off-point を機械的に `±1` としていましたが、これは整数・文字列長・配列長のような離散カウントのドメインにのみ妥当で、小数・金額・日付時刻のような連続量のドメインでは境界のすぐ外側を表しません。ドメインの最小意味単位（金額なら最小通貨単位、日時なら扱う精度の最小単位、浮動小数点なら丸め誤差を超える epsilon）を刻み幅にする、と一般化しました。

File: skills/test-design/references/classical-techniques.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+**例.** 割引率の決定（会員種別: 一般/プレミアム、購入額: 1万円未満/以上）:
+
+| 会員種別 | 購入額 | 割引率 |
+|---|---|---|
+| 一般 | 1万円未満 | 0% |
+| 一般 | 1万円以上 | 5% |
+| プレミアム | 1万円未満 | 10% |
+| プレミアム | 1万円以上 | 20% |
+
+条件が増えるほど `2^n` が爆発するため、無関係な条件の組が判明した時点で簡約するか、条件数が多い場合は §5 のペアワイズに切り替える。
```
[the remark this answers is not in the document]
Comment (kanade0404) [reply skills/test-design/references/classical-techniques.md]: Fixed in 99cc22a. Ready for re-review.

「条件数が多ければ §5 のペアワイズに切り替える」という記述は、3 条件以上の相互作用でのみ発生するアクション（デシジョンテーブルの特定の列）を、ペアワイズが 2 因子網羅しか保証しないために見落とすリスクがありました。ペアワイズへの切替は、デシジョンテーブルが要求する全アクションを最低 1 回ずつ網羅した後に限り、それ以外の独立した入力次元の間引きにのみ使う、と明記しました。

**Comment by kanade0404** [comment]
## Review Response Summary (2026-07-05 12:30 JST)

| Reviewer | Total | Fixed | Pushback | Deferred | Duplicate |
|---|---|---|---|---|---|
| @chatgpt-codex-connector | 8 | 7 | 0 | 1 | 0 |

### Deferred
- [https://github.com/kanade0404/skills/pull/33#discussion_r3524171039] `test-mutation-gate` ルーティング先が現時点で存在しない (SKILL.md 無し、master/feat/test-design 未マージ): 当該 skill 自体は別 PR (#28 "Add test-mutation-gate skill (Phase 1)", #35 がその上にスタック) として開発中で、そのマージ順は本 PR 単体の制御範囲外 → Tracked in #28

### Fixed (commit)
- [https://github.com/kanade0404/skills/pull/33#discussion_r3524171043] rulesync 生成物 (`.claude`/`.agents`) が未コミットで drift check が落ちる懸念 → `1e7727d` (本 PR 内で既に対応済み、本セッションで drift 無しを再確認)
- [https://github.com/kanade0404/skills/pull/33#discussion_r3524171047] fast-check (JS/TS) ユーザーを Python/Hypothesis 専用の `python.md` に誘導していた → `99cc22a`
- [https://github.com/kanade0404/skills/pull/33#discussion_r3524171050] Small サイズでの real-clock 例外が同 reference の hermetic/決定的入力生成の原則と矛盾 → `99cc22a`
- [https://github.com/kanade0404/skills/pull/33#discussion_r3524171053] defensive/wide contract の選択がテスト設計スキル自身の判断であるかのような記述で、design/software-design との境界・未決前提の規律と矛盾 → `99cc22a`
- [https://github.com/kanade0404/skills/pull/33#discussion_r3524171055] ペアワイズの水準列挙に無効クラスが混入し得た (§1 の独立テスト原則と矛盾) → `99cc22a`
- [https://github.com/kanade0404/skills/pull/33#discussion_r3524171057] BVA off-point の `±1` が小数・金額・日付時刻ドメインで境界の意味を持たない → `99cc22a`
- [https://github.com/kanade0404/skills/pull/33#discussion_r3524200703] デシジョンテーブル→ペアワイズ切替で 3 条件以上の相互作用アクションが網羅漏れになる懸念 → `99cc22a`

各スレッドへの返信は thread 内に投稿済み。レビュアーは `chatgpt-codex-connector` (Codex connector) のため vendor=human 扱いとし、全スレッドで `reply` のみ (resolve は発行していません)。
